File: scala/competitive/src/main/scala/leetcode/Question1414.scala
package leetcode

object Question1414 extends App {

  def findMinFibonacciNumbers(k: Int): Int = {
    import scala.collection.mutable
    val treeSet = mutable.TreeSet[Int]()(Ordering[Int].reverse)
    var x = 0
    var y = 1
    while (x <= k && y <=k) {
      val t = y
      y = x + y
      treeSet.add(y)
      x = t
    }
    var count = 0
    var i = k
    while (i > 0) {
      i = i - treeSet.from(i).to(1).firstKey
      count = count + 1
    }
    count
  }

//  println(findMinFibonacciNumbers(10000*10000*10))
  println(findMinFibonacciNumbers(19))
}

File: python/competitive/leetcode/Question1414.py
class Question1414:
    def findMinFibonacciNumbers(self, k: int) -> int:
        import collections
        import bisect

        # initializing dictionary
        test_dict = collections.OrderedDict()
        x = 0
        y = 1
        while x <= k and y <= k:
            t = y
            y = x + y
            test_dict[y] = y
            x = t
        test_dict = list(test_dict.keys())
        count = 0
        i = k
        while i > 0:
            i = i - test_dict[bisect.bisect_right(test_dict, i) - 1]
            count = count + 1
        return count


print(Question1414().findMinFibonacciNumbers(19))
